### PR TITLE
feature/v2

### PR DESCRIPTION
**Pull request by du-phan**
This PR adds new components, several enhancements and bug fixes.

**New components:**
* A recipe to compute drift between a deployed model and a dataset in the flow.
* A recipe to compute drift between two datasets in the flow.
* A custom metric that retrieve the most recent drift score in the resulted dataset from the two recipes above.

**Main enhancements:**
* Add support for regressions in the model view.
* Hide unnecessary information on model view when needed.
* Add support for all algorithms: for those that are not tree-based (and thus can not be used in v1), we use a surrogate model to get the feature importance information.
* Add `riskiest features` information allowing user to immediately have the list of features that they need to be careful about (ie. features that are drifted the most and are important in the deloyed model).
* Add support for partitioning.
* Add support for all types of train-test split (with/without cross-validation).

**Main bug fixes:**
* Fix bug with boolean dtype handling that leads to [mismatch prediction probability](https://app.clubhouse.io/dataiku/story/48162/prediction-proba-mismatch-between-public-api-and-scoring-recipe) and [weird categorical variable encoding](https://app.clubhouse.io/dataiku/story/47975/boolean-feature-handling-mismatch).   
* Fix bug with Date dtype and python 3.

### CODE REVIEW

File: plugin.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {
     "id": "model-drift",
-    "version": "1.0.0",
+    "version": "2.0.0",
     "meta": {
         "label": "Model drift monitoring",
-        "description": "Get insights on the applicability of a model by detecting when a new dataset differs from the one used for evaluation after training.",
-        "author": "Dataiku (Léo Dreyfus-Schmidt & Du Phan)",
+        "description": "Get insights on the degree of drift between two (training) datasets.",
```
Comment (cperdigou) [inline plugin.json:6]: Maybe "Get insights on feature drift between two (training) datasets."? Or, "on how much feature drift has happened"

File: python-lib/dku_data_drift/dataframe_helpers.py [hunk 2]
```diff
@@ -1,22 +1,56 @@
-# coding: utf-8
-import numpy as np
+#!/usr/bin/python
```
Comment (cperdigou) [inline python-lib/dku_data_drift/dataframe_helpers.py:1]: Why do you need a shebang??

File: python-probes/drift-score/probe.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+    "meta" : {
+        "label": "Most recent drift score",
+        "description": "",
+        "icon": "icon-puzzle-piece"
```
Comment (cperdigou) [inline python-probes/drift-score/probe.json:5]: my not the anchor?

File: resource/compute_model_id_choice.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+"""
+Allow dynamic select of the model id in the model recipe.
+"""
+import dataiku
+import datetime
+
+
+def process_timestamp(timestamp):
```
Comment (cperdigou) [inline resource/compute_model_id_choice.py:8]: dup from lib

File: webapps/model-drift-view/backend.py [hunk 5]
```diff
@@ -8,6 +8,8 @@
 from model_metadata import get_model_handler
 logger = logging.getLogger(__name__)
 
+MAX_NUM_ROW = 100000 # heuristic choice
```
Comment (MayeulRousselet) [inline webapps/model-drift-view/backend.py:11]: Constantify

File: webapps/model-drift-view/backend.py [hunk 6]
```diff
@@ -8,6 +8,8 @@
...
+
 
 def convert(o):
```
Comment (MayeulRousselet) [inline webapps/model-drift-view/backend.py:14]: Give a more explicit name: convert to what?

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    raise ValueError('Without the partition column, dataset is empty.')
+
+# Retrieve input model
+logger.info("Retrieve the input model")
+input_names = get_input_names_for_role('model')
+model = dataiku.Model(input_names[0])
+model_id = model.get_id()
+
+# Retrieve the version id of the model (dynamic dropdown selection)
+use_active_version = get_recipe_config().get('use_active_version')
```
Comment (MayeulRousselet) [inline custom-recipes/compare-dataset-with-model/recipe.py:43]: Could be a method called get_active_version

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+drifter = DriftAnalyzer(prediction_type=None)
+target = model_accessor.get_target_variable()
+drifter.fit(new_df, model_accessor=model_accessor)
+
+# Write the drift score and metrics
+
+model_train_date = get_train_date(model_id, version_id)
+
+timestamp = datetime.datetime.now()
+new_df = pd.DataFrame({'timestamp': [timestamp],
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        new_df['fugacity'] = json.dumps(fugacity)
+        new_df['fugacity_relative_change'] = json.dumps(fugacity_relative_change)
+        column_description_dict['fugacity'] = 'Proportion of samples predicted (in %) in each class when scoring on both the original test and the new input dataset.'
+        column_description_dict['fugacity_relative_change'] = 'Relative change (in %) in each class with respect to the original fugacity value.\n\nFormula: 100*(new_fugacity - original_fugacity)/original_fugacity'
+    else: # regression
+        fugacity, fugacity_relative_change, bin_description = drifter.get_regression_fugacity()
+        new_df['fugacity'] = json.dumps(fugacity)
+        new_df['fugacity_relative_change'] = json.dumps(fugacity_relative_change)
+        proper_bin_description = '\n'.join(['Decile {0}: {1}'.format(bin_index, bin_desc) for bin_index, bin_desc in enumerate(bin_description)])
+        fugacity_def = 'Proportion of samples predicted (in %) in each decile when scoring on both the original test and the new input dataset.\n\n'
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
+import dataiku
+import pandas as pd
+import json
+from dataiku.customrecipe import *
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+from dku_data_drift.drift_analyzer import DriftAnalyzer
+from dku_data_drift.model_accessor import ModelAccessor
+from dku_data_drift.dataset_helpers import get_partitioning_columns
+
+from model_metadata import get_model_handler
+
+import datetime
+import logging
+
+MAX_NUM_ROW = 100000 # heuristic choice
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in (suspens)

File: custom-recipes/compare-datasets/recipe.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+import json
+from dku_tools import set_column_description
+from dataiku.customrecipe import *
+from dku_data_drift.drift_analyzer import DriftAnalyzer
+from dku_data_drift.dataset_helpers import get_partitioning_columns
+
+import datetime
+import logging
+
+MAX_NUM_ROW = 100000 # heuristic choice
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-datasets/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-datasets/recipe.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
+import dataiku
+import pandas as pd
+import json
+from dku_tools import set_column_description
+from dataiku.customrecipe import *
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-datasets/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+    drift_feature_importance = drifter.get_drift_feature_importance()
+    original_feature_importance = drifter.get_original_feature_importance()
+
+    if 'riskiest_features' in metric_list:
+        riskiest_feature = drifter.get_riskiest_features(drift_feature_importance, original_feature_importance)
+        new_df['riskiest_features'] = json.dumps(riskiest_feature)
+        column_description_dict['riskiest_features'] = 'If the drift score is medium/high (above 0.1), we recommend you to check those features.\nA feature is considered risky if it is both in the top 40% of the most drifted features as well as the top 40% most important features in the original model.'
+    feat_dict = {}
+    for feat, feat_info in drift_feature_importance[:20].iterrows():
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+output_datasets = [dataiku.Dataset(name) for name in output_names]
+output_dataset = output_datasets[0]
+
+# Access the model
+model_handler = get_model_handler(model=model, version_id=version_id)
+model_accessor = ModelAccessor(model_handler)
+
+# Analyze the drift
+drifter = DriftAnalyzer(prediction_type=None)
+target = model_accessor.get_target_variable()
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1

File: custom-recipes/compare-datasets/recipe.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+if 'drift_score' in metric_list:
+    drift_score = drifter.get_drift_score()
+    new_df['drift_score'] = [drift_score]
+    column_description_dict['drift_score'] = 'The drift score (between 0 and 1) is low (under 0.1) if the new dataset and the original dataset are indistinguishable.'
+
+if 'feature_importance' in metric_list:
+    feature_importance = drifter.get_drift_feature_importance()
+    feat_dict = {}
+    for feat, feat_info in feature_importance[:10].iterrows():
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-datasets/recipe.py]: Done in 5d4e6f1

File: python-lib/dku_data_drift/model_accessor.py [hunk 17]
```diff
@@ -1,46 +1,91 @@
...
-            feature_importance.append({
-                'feature': feature_name,
-                'importance': 100*feat_importance/sum(feature_importances)
-            })
-        
-        dfx = pd.DataFrame(feature_importance).sort_values(by='importance', ascending=False).reset_index(drop=True)
-        dfx['cumulative_importance'] = dfx['importance'].cumsum()
-        dfx_top = dfx.loc[dfx['cumulative_importance'] <= cumulative_percentage_threshold]
-        return dfx_top.rename_axis('rank').reset_index().set_index('feature')
+    def get_feature_importance(self, cumulative_percentage_threshold=95):
```
[the remark this answers is not in the document]
Comment (du-phan) [reply python-lib/dku_data_drift/model_accessor.py]: Done in 5d4e6f1

File: webapps/model-drift-view/backend.py (hunk 5 above)
Comment (du-phan) [reply webapps/model-drift-view/backend.py]: Done in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+#fix the column order
+new_df = new_df[['timestamp', 'model_id', 'version_id', 'train_date']]
+
+column_description_dict = {}
+if 'drift_score' in metric_list:
+    drift_score = drifter.get_drift_score()
+    new_df['drift_score'] = [drift_score]
+    column_description_dict['drift_score'] = 'The drift score (between 0 and 1) is low (udner 0.1) if the new dataset and the original dataset are indistinguishable.'
+
+if 'fugacity' in metric_list:
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1

File: python-lib/dku_data_drift/dataframe_helpers.py (hunk 2 above)
Comment (du-phan) [reply python-lib/dku_data_drift/dataframe_helpers.py:1]: Legacy codes ... clean up in 5d4e6f1

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+new_df = pd.DataFrame({'timestamp': [timestamp],
+                       'model_id': [model_id],
+                       'version_id': [version_id],
+                       'train_date': [model_train_date]})
+
+#fix the column order
+new_df = new_df[['timestamp', 'model_id', 'version_id', 'train_date']]
+
+column_description_dict = {}
+if 'drift_score' in metric_list:
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: I'm wondering if restrict to only py3 is a good idea, for this specific plugin.
The main problem is that for the plugin to work with a specific saved model deployed on the flow, the 2 must have same python version. People usually use the default dku code-env for their ML task, which I feel like most of the time is still python 2.

File: python-probes/drift-score/probe.py [hunk 20]
```diff
@@ -0,0 +1,7 @@
+import numpy as np
+
+def process(dataset, partition_id):
+    df = dataset.get_dataframe()
+    most_recent_drift_score = df[df['timestamp'] == np.max(df['timestamp'])]['drift_score'].values[0]
```
[the remark this answers is not in the document]
Comment (du-phan) [reply python-probes/drift-score/probe.py]: Done in 8ea647c

File: python-lib/dku_data_drift/model_tools.py [hunk 21]
```diff
@@ -59,3 +66,73 @@ def A(i, j):
...
+
+    def fit(self, df, target):
+        preprocessor = Preprocessor(df, target)
+        train, test = preprocessor.get_processed_train_test()
+        train_X = train.drop(target, axis=1)
+        train_Y = train[target]
+        self.clf.fit(train_X, train_Y)
+        self.feature_names = train_X.columns
+
+    def get_feature_importance(self, cumulative_percentage_threshold=95):
```
[the remark this answers is not in the document]
Comment (du-phan) [reply python-lib/dku_data_drift/model_tools.py]: I was hoping nobody would notice this ahaha

File: python-lib/dku_data_drift/model_tools.py [hunk 22]
```diff
@@ -59,3 +66,73 @@ def A(i, j):
...
+    Y_plot = [v if not np.isnan(v) else 0 for v in np.exp(kde.score_samples(X_plot))]
+    return list(zip(X_plot.ravel(), Y_plot))
+
+class SurrogateModel:
+
+    def __init__(self, prediction_type):
+        self.feature_names = None
+        self.target = None
+        self.prediction_type = prediction_type
+        #TODO should we define some params of RF to avoid long computation ?
```
[the remark this answers is not in the document]
Comment (du-phan) [reply python-lib/dku_data_drift/model_tools.py:112]: Done in 604ec98

File: python-lib/dku_data_drift/model_tools.py [hunk 23]
```diff
@@ -59,3 +66,73 @@ def A(i, j):
...
+        h = 0.06
+    if len(np.unique(data)) == 1:
+        sample_weight = None
+    # Definition of the support of the estimate
+    X_plot = np.linspace(min_support, max_support, 500, dtype=float)[:, np.newaxis]
+    kde = KernelDensity(kernel='gaussian', bandwidth=h).fit(data.reshape(-1, 1), sample_weight=sample_weight)
+    Y_plot = [v if not np.isnan(v) else 0 for v in np.exp(kde.score_samples(X_plot))]
+    return list(zip(X_plot.ravel(), Y_plot))
+
+class SurrogateModel:
```
[the remark this answers is not in the document]
Comment (du-phan) [reply python-lib/dku_data_drift/model_tools.py]: Done in 604ec98

File: webapps/model-drift-view/backend.py (hunk 6 above)
Comment (du-phan) [reply webapps/model-drift-view/backend.py]: Done in 1dfe086

File: python-lib/dku_data_drift/model_accessor.py [hunk 24]
```diff
@@ -1,46 +1,91 @@
...
+        """
```
[the remark this answers is not in the document]
Comment (du-phan) [reply python-lib/dku_data_drift/model_accessor.py:56]: Done in 1dfe086

File: code-env/python/spec/requirements.txt [hunk 25]
```diff
@@ -2,4 +2,7 @@ flask
 scikit-learn==0.20.2
 scipy==1.1.0
 xgboost==0.81
-future
\ No newline at end of file
+pandas==0.23.4
+numpy==1.16.6
+future
+joblib
```
[the remark this answers is not in the document]
Comment (du-phan) [reply code-env/python/spec/requirements.txt]: Done in 918db57

File: python-probes/drift-score/probe.json (hunk 3 above)
Comment (du-phan) [reply python-probes/drift-score/probe.json]: done

File: custom-recipes/compare-dataset-with-model/recipe.json [hunk 26]
```diff
@@ -0,0 +1,76 @@
+{
+    "meta": {
+        "label": "Compute feature drift of a deployed model",
+        "description": "Measure a potential drift between the original training data of a model and a new dataset. Support regression and classification",
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.json]: Done in f35450b

File: custom-recipes/compare-dataset-with-model/recipe.json [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+            "required": true,
+            "acceptsDataset": true
+        }
+    ],
+
+    "outputRoles" : [
+        {
+            "name": "main_output",
+            "label": "Metrics dataset",
+            "description": "Dataset storing drift metrics",
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.json]: Done in f35450b

File: custom-recipes/compare-datasets/recipe.json [hunk 28]
```diff
@@ -0,0 +1,61 @@
+{
+    "meta": {
+        "label": "Compute drift between two datasets",
+        "description": "Measure a potential difference between two datasets having the same schema.",
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-datasets/recipe.json]: Done in f35450b

File: plugin.json (hunk 1 above)
Comment (du-phan) [reply plugin.json]: Done in f35450b

File: resource/compute_model_id_choice.py (hunk 4 above)
Comment (du-phan) [reply resource/compute_model_id_choice.py]: Done in 3b7b448

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+    column_description_dict['drift_score'] = 'The drift score (between 0 and 1) is low (udner 0.1) if the new dataset and the original dataset are indistinguishable.'
+
+if 'fugacity' in metric_list:
+    if drifter.get_prediction_type() == 'CLASSIFICATION':
+        fugacity, fugacity_relative_change = drifter.get_classification_fugacity()
+        new_df['fugacity'] = json.dumps(fugacity)
+        new_df['fugacity_relative_change'] = json.dumps(fugacity_relative_change)
+        column_description_dict['fugacity'] = 'Proportion of samples predicted (in %) in each class when scoring on both the original test and the new input dataset.'
+        column_description_dict['fugacity_relative_change'] = 'Relative change (in %) in each class with respect to the original fugacity value.\n\nFormula: 100*(new_fugacity - original_fugacity)/original_fugacity'
+    else: # regression
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in f8b15d2

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+
+    original_feature_importance = drifter.get_original_feature_importance()
+    feat_dict = {}
+    for feat, feat_info in original_feature_importance[:20].iterrows():
+        feat_dict[feat] = round(feat_info.get('importance'), 2)
+    new_df['most_important_features_in_deployed_model'] = [json.dumps(feat_dict)]
+    column_description_dict['most_important_features_in_deployed_model'] = 'Most important features in the deployed model, with their % of importance (max 20 features).'
+
+elif 'riskiest_features' in metric_list:
+    riskiest_feature = drifter.get_riskiest_features()
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py:95]: Done

File: custom-recipes/compare-dataset-with-model/recipe.py (hunk 7 above)
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in f8b15d2

File: python-lib/dku_data_drift/model_tools.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (du-phan) [reply python-lib/dku_data_drift/model_tools.py]: Done in edb5b64

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 31]
```diff
@@ -0,0 +1,140 @@
...
+                       'version_id': [version_id],
+                       'train_date': [model_train_date]})
+
+#fix the column order
+new_df = new_df[['timestamp', 'model_id', 'version_id', 'train_date']]
+
+column_description_dict = {}
+if 'drift_score' in metric_list:
+    drift_score = drifter.get_drift_score()
+    new_df['drift_score'] = [drift_score]
```
[the remark this answers is not in the document]
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5d4e6f1 
For now I just use constant for py2/py3 compatibility

**Comment by du-phan** [comment]
@MayeulRousselet I've addressed all your comments in the recent commits. Back to you thank you 🙏

File: custom-recipes/compare-dataset-with-model/recipe.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (MayeulRousselet) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Good comment, you're right!

File: python-lib/dku_data_drift/model_accessor.py [hunk 32]
```diff
@@ -1,46 +1,91 @@
...
+        Wrap the prediction type accessor of the model
+        """
+        if ModelDriftConstants.CLASSIFICATION_TYPE in self.model_handler.get_prediction_type():
+            return ModelDriftConstants.CLASSIFICATION_TYPE
+        elif ModelDriftConstants.REGRRSSION_TYPE in self.model_handler.get_prediction_type():
+            return ModelDriftConstants.REGRRSSION_TYPE
+        else:
+            return ModelDriftConstants.CLUSTERING_TYPE
+
     def check(self):
```
Comment (MayeulRousselet) [inline python-lib/dku_data_drift/model_accessor.py:29]: Don't seems to be called, either call or remove

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+                       ModelDriftConstants.MODEL_ID: [model.get_id()],
+                       ModelDriftConstants.VERSION_ID: [version_id],
+                       ModelDriftConstants.TRAIN_DATE: [model_train_date]})
+
+#fix the column order
+new_df = new_df[[ModelDriftConstants.TIMESTAMP, ModelDriftConstants.MODEL_ID, ModelDriftConstants.VERSION_ID, ModelDriftConstants.TRAIN_DATE]]
+
+column_description_dict = {}
+if ModelDriftConstants.DRIFT_SCORE in metric_list:
+    drift_score = drifter.get_drift_score()
```
Comment (MayeulRousselet) [inline custom-recipes/compare-dataset-with-model/recipe.py:50]: Could be a method "extract_drift_score" for example. 
And could be called here and in the compare dataset recipe as they both do the same thing.

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 34]
```diff
@@ -0,0 +1,102 @@
...
+        column_description_dict[ModelDriftConstants.RISKIEST_FEATURES] = ModelDriftConstants.RISKIEST_FEATURES_DEFINITION
+    feat_dict = {}
+    for feat, feat_info in drift_feature_importance[:ModelDriftConstants.NUMBER_OF_DRIFTED_FEATURES].iterrows():
+        feat_dict[feat] = round(feat_info.get(ModelDriftConstants.IMPORTANCE), 2)
+    new_df[ModelDriftConstants.MOST_DRIFTED_FEATURES] = [json.dumps(feat_dict)]
+    column_description_dict[ModelDriftConstants.MOST_DRIFTED_FEATURES] = ModelDriftConstants.MOST_DRIFTED_FEATURES_DEFINITION
+
+    original_feature_importance = drifter.get_original_feature_importance()
+    feat_dict = {}
+    for feat, feat_info in original_feature_importance[:ModelDriftConstants.NUMBER_OF_DRIFTED_FEATURES].iterrows():
```
Comment (MayeulRousselet) [inline custom-recipes/compare-dataset-with-model/recipe.py:89]: I think we could factorize with what it is done in compare_dataset

File: custom-recipes/compare-dataset-with-model/recipe.py [hunk 35]
```diff
@@ -0,0 +1,102 @@
...
+        new_df[ModelDriftConstants.FUGACITY] = json.dumps(fugacity)
+        new_df[ModelDriftConstants.FUGACITY_RELATIVE_CHANGE] = json.dumps(fugacity_relative_change)
+        proper_bin_description = '\n'.join(['Decile {0}: {1}'.format(bin_index, bin_desc) for bin_index, bin_desc in enumerate(bin_description)])
+        column_description_dict[ModelDriftConstants.FUGACITY] = ModelDriftConstants.FUGACITY_REGRESSION_DEFINITION + proper_bin_description
+        column_description_dict[ModelDriftConstants.FUGACITY_RELATIVE_CHANGE] = ModelDriftConstants.FUGACITY_RELATIVE_CHANGE_REGRESSION_DEFINITION + proper_bin_description
+    else:
+        raise ValueError('Unsupported prediction type: {0}'.format(drifter.get_prediction_type()))
+
+
+if ModelDriftConstants.FEATURE_IMPORTANCE in metric_list:
```
Comment (MayeulRousselet) [inline custom-recipes/compare-dataset-with-model/recipe.py:72]: same here

File: python-lib/dku_tools.py [hunk 36]
```diff
@@ -0,0 +1,87 @@
...
+def process_timestamp(timestamp):
+    """
+    Convert the timestamp to str date
+    :param timestamp:
+    :return:
+    """
+    return str(datetime.datetime.fromtimestamp(timestamp / 1000))
+
+
+def get_train_date(model_version, version_id):
```
Comment (MayeulRousselet) [inline python-lib/dku_tools.py:16]: Shouldn't it be in model_metadata?

File: python-lib/dku_data_drift/model_accessor.py (hunk 32 above)
Comment (du-phan) [reply python-lib/dku_data_drift/model_accessor.py]: Done in 38e108e

File: custom-recipes/compare-dataset-with-model/recipe.py (hunk 33 above)
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5a26c7d

File: custom-recipes/compare-dataset-with-model/recipe.py (hunk 34 above)
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5a26c7d

File: custom-recipes/compare-dataset-with-model/recipe.py (hunk 35 above)
Comment (du-phan) [reply custom-recipes/compare-dataset-with-model/recipe.py]: Done in 5a26c7d

File: python-lib/dku_tools.py (hunk 36 above)
Comment (du-phan) [reply python-lib/dku_tools.py]: Indeed! move it there in d5f407a

**Comment by du-phan** [comment]
@MayeulRousselet I finally managed to finish this round, back to you 🙏

**Pull request merged**